### PR TITLE
Add vuejs blogpost

### PR DESCRIPTION
**Pull request by mooreds**
This is a vuejs/oauth blog post

### CODE REVIEW

File: site/_posts/2020-07-22-securely-implement-oauth-vuejs.md [hunk 1]
```diff
@@ -0,0 +1,1217 @@
...
+
+{% include _image.liquid src="/assets/img/blogs/oauth-vuejs/initial-error.png" alt="An error screen on a login attempt" class="img-fluid" figure=false %}
+
+If you review `login.js` above, you will find that `REDIRECT_URI` is set to `http://localhost:9000/oauth-callback` in our `.env` file but we haven’t actually created that route yet. So this error makes sense. We're actually logged in because we signed into the FusionAuth Dashboard from our previous configuration. 
+
+If you were using a new browser or an incognito window, you might see the login screen instead:
+
+{% include _image.liquid src="/assets/img/blogs/oauth-vuejs/default-login-screen.png" alt="The default login screen." class="img-fluid" figure=false %}
+
+## OAuth Callback
```
Comment (dpolstra) [inline site/_posts/2020-07-22-securely-implement-oauth-vuejs.md:662]: Creating an OAuth Callback for Authorization Code Flow

File: site/_posts/2020-07-22-securely-implement-oauth-vuejs.md [hunk 2]
```diff
@@ -0,0 +1,1217 @@
...
+  token_type: 'Bearer',
+  userId: 'abdee025-fa3c-4ce2-b6af-d0931cfb4cea'
+}
+```
+
+You can see what an `access_token` looks like. The axios request ends with a catch block to handle any error that we may encounter.
+
+Now, head over to http://localhost:9000/login. if everything goes well, you will end up on your Vue application homepage because that is what we have set in `redirect_uri` and can see the response in the console (the terminal where you are running your server).
+
+## Logout
```
Comment (dpolstra) [inline site/_posts/2020-07-22-securely-implement-oauth-vuejs.md:825]: Adding a Logout Route

File: site/_posts/2020-07-22-securely-implement-oauth-vuejs.md [hunk 3]
```diff
@@ -0,0 +1,1213 @@
...
+  }
+});
+module.exports = router;
+```
+
+Let’s discuss this step by step. 
+
+First, we check if an `access_token` is present and then make a **POST** request to `oauth2/introspect` endpoint which requires Client Id and the token. Like the `oauth2/token` endpoint, this endpoint expects form-encoded data, so we are using the `query-string` library.
+
+When this request is successful, we get a response object which contains a `data` property that has the information we need.
```
Comment (voidmain) [inline site/_posts/2020-07-22-securely-implement-oauth-vuejs.md:939]: I don't believe there is a `data` property in the introspect response.

File: site/_posts/2020-07-22-securely-implement-oauth-vuejs.md [hunk 4]
```diff
@@ -0,0 +1,1213 @@
...
+  "id": "c756e203-ea1f-491e-9446-b70ed4eecc17",
+  "insertInstant": 1594898302209,
+  "lastLoginInstant": 1594900452281,
+  "username": "ashu",
+  "usernameStatus": "ACTIVE",
+  "verified": true
+}
+```
+
+The API key that we are passing in headers is not part of OAuth standard. You need it to call non-standard endpoints such as the [User Registration API](https://fusionauth.io/docs/v1/tech/apis/registrations#retrieve-a-user-registration). We added this to show how you will use the API Key (in `Authorization` header) if you decide to explore more endpoints which require the API key.
```
Comment (voidmain) [inline site/_posts/2020-07-22-securely-implement-oauth-vuejs.md:975]: "in headers" should be "in the `Authorization` HTTP header"

File: site/_posts/2020-07-22-securely-implement-oauth-vuejs.md [hunk 5]
```diff
@@ -0,0 +1,1213 @@
...
+
+## Showing user data
+
+We can now access user's information that was stored in FusionAuth. The next step is to display that data. In our `App.vue` we need to modify `mounted()`, since this time we are getting a response object that contains data from two endpoints.
```
Comment (voidmain) [inline site/_posts/2020-07-22-securely-implement-oauth-vuejs.md:979]: "the user's information"?

File: site/_posts/2020-07-22-securely-implement-oauth-vuejs.md [hunk 6]
```diff
@@ -0,0 +1,1213 @@
...
+      console.error(err);
+    });
+
+});
+module.exports = router;
+```
+
+To ensure that we are updating the user currently logged in, we extract the token from our server which is done by making a **POST** request to the `oauth/introspect` endpoint; this is similar to what we did in the `user` route. 
+
+Once this request is successful, we make a **PATCH** request to `/api/user/registration` API. If you go through the [User Registration docs](https://fusionauth.io/docs/v1/tech/apis/registrations#update-a-user-registration), you will find that this API accepts both **PUT** and **PATCH** requests but here we are using **PATCH** since we only want to update a single part of the resource and **PATCH** will merge the request parameters into the existing object.
```
Comment (voidmain) [inline site/_posts/2020-07-22-securely-implement-oauth-vuejs.md:1106]: I'd change "resource" to "object"

File: site/_posts/2020-07-22-securely-implement-oauth-vuejs.md [hunk 7]
```diff
@@ -0,0 +1,1213 @@
...
+
+});
+module.exports = router;
+```
+
+To ensure that we are updating the user currently logged in, we extract the token from our server which is done by making a **POST** request to the `oauth/introspect` endpoint; this is similar to what we did in the `user` route. 
+
+Once this request is successful, we make a **PATCH** request to `/api/user/registration` API. If you go through the [User Registration docs](https://fusionauth.io/docs/v1/tech/apis/registrations#update-a-user-registration), you will find that this API accepts both **PUT** and **PATCH** requests but here we are using **PATCH** since we only want to update a single part of the resource and **PATCH** will merge the request parameters into the existing object. 
+
+The `data` to send is stored inside the `registration` object which takes it value from `req.body`. This `registration` represents a user's association with an application. The `data` attribute allows us to store arbitrary data related to a user's registration in an application. We will be sending user data from our Vue app to the Express server via JSON in the body of a **POST** HTTP message.
```
Comment (voidmain) [inline site/_posts/2020-07-22-securely-implement-oauth-vuejs.md:1108]: "takes it value" should probably be "takes its value". 

Also, this says **POST** but aren't we doing **PATCH**?

File: site/_posts/2020-07-22-securely-implement-oauth-vuejs.md [hunk 8]
```diff
@@ -0,0 +1,1213 @@
...
+  margin-top: 5px;
+}
+</style>
+```
+
+One of the cool features of Vue is that by using `v-model="userData"` and initializing `userData=""` in the `data()` function, it creates a two-way data binding between `textarea` and `userData`.
+
+We can now access whatever we type in `textarea` in `userData`. You can [read more about it here](https://vuejs.org/v2/guide/forms.html).
+
+Then add this component to `App.vue` . It does not make sense to show this component when the user is not logged in. To hide it, just add `v-if="email"` with this component. Like some of our other components it will check to see if `email` is present or not, i.e., if user is logged in or not.
```
Comment (voidmain) [inline site/_posts/2020-07-22-securely-implement-oauth-vuejs.md:1153]: The last sentence needs work. It needs a comma after the "like" clause and the "i.e." part should probably be in parenthesis and might not need any of the commas.

File: site/_posts/2020-07-22-securely-implement-oauth-vuejs.md [hunk 9]
```diff
@@ -0,0 +1,1213 @@
...
+//...
+```
+
+In the above function, we use `fetch()` to **POST** JSON-encoded data. If you are familiar with `fetch()` API, you will see that this is a simple **POST** request using `fetch()`, nothing fancy. You can [read more about it here](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch).
+
+Once we have sent `userData` to our server, we empty the `textarea` using `this.userData=''`, since it is a two-way binding.
+
+And now to bind this function to `submit` event we will add the following to the `form` tag.
+
+```html
```
Comment (voidmain) [inline site/_posts/2020-07-22-securely-implement-oauth-vuejs.md:1189]: This code type should use whatever language supports the Vue syntax. Otherwise, it renders as an error with the "@" attribute.

File: site/_posts/2020-07-22-securely-implement-oauth-vuejs.md [hunk 10]
```diff
@@ -0,0 +1,1213 @@
...
+
+Using `.prevent` stops the page from reloading whenever the Submit button is clicked. Here is how our application looks now:
+
+{% include _image.liquid src="/assets/img/blogs/oauth-vuejs/signed-in-with-text-area.png" alt="The application when logged in with the text area for updating the data." class="img-fluid" figure=false %}
+
+Now go to your Vue app and type some message in the `textarea` and click the Submit button. You can now see that message under **User data** in **Users** tab in the FusionAuth dashboard. 
+
+## Conclusion
+
+Congrats, you have built a Vue app that can log in, log out, and modify user data. This article provides a foundation for implementing OAuth using FusionAuth, and there are still so many features, components, and routes that you can add.
```
Comment (voidmain) [inline site/_posts/2020-07-22-securely-implement-oauth-vuejs.md:1203]: I'd reword "still so many" to "a bunch of other".

File: site/_posts/2020-07-22-securely-implement-oauth-vuejs.md [hunk 11]
```diff
@@ -0,0 +1,1213 @@
...
+
+{% include _image.liquid src="/assets/img/blogs/oauth-vuejs/signed-in-with-text-area.png" alt="The application when logged in with the text area for updating the data." class="img-fluid" figure=false %}
+
+Now go to your Vue app and type some message in the `textarea` and click the Submit button. You can now see that message under **User data** in **Users** tab in the FusionAuth dashboard. 
+
+## Conclusion
+
+Congrats, you have built a Vue app that can log in, log out, and modify user data. This article provides a foundation for implementing OAuth using FusionAuth, and there are still so many features, components, and routes that you can add. 
+
+Again, here's [the code](https://github.com/fusionauth/fusionauth-example-vue) that you can fork and look at.
```
Comment (voidmain) [inline site/_posts/2020-07-22-securely-implement-oauth-vuejs.md:1205]: I'd change "look at" to "experiment with" or "play around with".

File: site/_posts/2020-07-22-securely-implement-oauth-vuejs.md [hunk 12]
```diff
@@ -0,0 +1,1213 @@
...
+
+Here are a few ideas of what you can do next:
+
+- [Register Users from the App itself.](https://fusionauth.io/docs/v1/tech/apis/registrations#create-a-user-and-registration-combined)
+- Secure your server using middlewares like [Helmet,](https://www.npmjs.com/package/helmet)
```
Comment (voidmain) [inline site/_posts/2020-07-22-securely-implement-oauth-vuejs.md:1210]: Is the comma needed at the end of the line?

File: site/_posts/2020-07-22-securely-implement-oauth-vuejs.md [hunk 13]
```diff
@@ -0,0 +1,1213 @@
...
+- [query-string](https://www.npmjs.com/package/query-string) -  to stringify form data that we send using `axios`.
+
+As we have installed `nodemon`, to use it inside `package.json` add the following scripts:
+
+```json
+...
+"scripts": {
+  "start": "node index.js",
+  "dev": "nodemon index.js"
+},
```
[the remark this answers is not in the document]
Comment (mooreds) [reply site/_posts/2020-07-22-securely-implement-oauth-vuejs.md:193]: There's not a great answer for this, as we've discussed. I ended up going with the javascript formatter for any JSON that I wanted to excerpt.

File: site/_posts/2020-07-22-securely-implement-oauth-vuejs.md (hunk 7 above)
Comment (mooreds) [reply site/_posts/2020-07-22-securely-implement-oauth-vuejs.md:1108]: The Vue app is `POST`ing to the express app, which is then `PATCH`ing to the FusionAuth app.

Will rework to be less confusing.

File: site/_posts/2020-07-22-securely-implement-oauth-vuejs.md (hunk 9 above)
Comment (mooreds) [reply site/_posts/2020-07-22-securely-implement-oauth-vuejs.md:1189]: Yup, it's `vue`. Here's the list of all the formatters supported by `rouge`, the jekyll code highlighter: https://github.com/rouge-ruby/rouge/tree/master/lib/rouge/lexers

File: site/_posts/2020-07-22-securely-implement-oauth-vuejs.md [hunk 14]
```diff
@@ -0,0 +1,1217 @@
...
+
+```
+user:{
+  email: 'richard@fusionauth.io'
+}
+```
+
+{% include _image.liquid src="/assets/img/blogs/oauth-vuejs/welcome-screen-hardcoded-richard.png" alt="The welcome screen with Richard's hardcoded data." class="img-fluid" figure=false %}
+
+## Sending Data from FusionAuth
```
[the remark this answers is not in the document]
Comment (mooreds) [reply site/_posts/2020-07-22-securely-implement-oauth-vuejs.md]: Sorry, that doesn't make sense :)
